pkg/virt-controller/watch/workload-updater/workload-updater.go

```diff
@@ -0,0 +1,571 @@
+
+	ts, ok := c.reconcileThrottleMap[key]
+	if !ok {
+		c.reconcileThrottleMap[key] = now.Add(time.Duration(c.throttleIntervalSeconds) * time.Second)
+	} else if now.Before(ts) {
+		c.queue.AddAfter(key, time.Duration(c.throttleIntervalSeconds)*time.Second)
+		return nil
+	}
+
+	kv := obj.(*virtv1.KubeVirt)
+
+	// don't update workloads unless the infra is completely deployed and not updating
+	if kv.Status.Phase != virtv1.KubeVirtPhaseDeployed {
+		return nil
+	} else if kv.Status.ObservedDeploymentID != kv.Status.TargetDeploymentID {
+		return nil
+	}
+
+	return c.sync(kv)
+}
+
+func (c *WorkloadUpdateController) sync(kv *virtv1.KubeVirt) error {
+
+	data := c.getUpdateData(kv)
+
+	key, err := controller.KeyFunc(kv)
+	if err != nil {
+		return err
+	}
+
+	outdatedVirtualMachineInstanceWorkloads.Set(float64(len(data.allOutdatedVMIs)))
+
+	// update outdated workload count on kv
+	if kv.Status.OutdatedVirtualMachineInstanceWorkloads == nil || *kv.Status.OutdatedVirtualMachineInstanceWorkloads != len(data.allOutdatedVMIs) {
+		l := len(data.allOutdatedVMIs)
+		kvCopy := kv.DeepCopy()
+		kvCopy.Status.OutdatedVirtualMachineInstanceWorkloads = &l
+
+		oldJson, err := json.Marshal(kv.Status.OutdatedVirtualMachineInstanceWorkloads)
+		if err != nil {
+			return err
+		}
+
+		newJson, err := json.Marshal(kvCopy.Status.OutdatedVirtualMachineInstanceWorkloads)
+		if err != nil {
+			return err
+		}
+
+		patch := ""
+		if kv.Status.OutdatedVirtualMachineInstanceWorkloads == nil {
+			update := fmt.Sprintf(`{ "op": "add", "path": "/status/outdatedVirtualMachineInstanceWorkloads", "value": %s}`, string(newJson))
+			patch = fmt.Sprintf("[%s]", update)
+		} else {
+			test := fmt.Sprintf(`{ "op": "test", "path": "/status/outdatedVirtualMachineInstanceWorkloads", "value": %s}`, string(oldJson))
+			update := fmt.Sprintf(`{ "op": "replace", "path": "/status/outdatedVirtualMachineInstanceWorkloads", "value": %s}`, string(newJson))
+			patch = fmt.Sprintf("[%s, %s]", test, update)
+		}
+
+		err = c.statusUpdater.PatchStatus(kv, types.JSONPatchType, []byte(patch))
+		if err != nil {
+			return fmt.Errorf("unable to patch kubevirt obj status to update the outdatedVirtualMachineInstanceWorkloads valued: %v", err)
+		}
+	}
+
+	// Rather than enqueing based on VMI activity, we keep periodically poping the loop
+	// until all VMIs are updated. Watching all VMI activity is chatty for this controller
+	// when we don't need to be that efficent in how quickly the updates are being processed.
+	if len(data.evictOutdatedVMIs) != 0 || len(data.migratableOutdatedVMIs) != 0 {
+		c.queue.AddAfter(key, periodicReEnqueueIntervalSeconds)
+	}
+
+	// Randomizes list so we don't always re-attempt the same vmis in
+	// the event that some are having difficulty being relocated
+	rand.Shuffle(len(data.migratableOutdatedVMIs), func(i, j int) {
+		data.migratableOutdatedVMIs[i], data.migratableOutdatedVMIs[j] = data.migratableOutdatedVMIs[j], data.migratableOutdatedVMIs[i]
+	})
```

Comment:
I guess we could just form this list of `migratableOutdatedVMIs` in a way that VMIs that have a failed migration in its migrationStatus will be last, but I'm fine with his approach too